### PR TITLE
e2e testing using minikube

### PR DESCRIPTION
**Pull request by vishh**
This PR adds a new make rule `make test-e2e-minikube` that will run portable e2e tests against a local minikube cluster. 
This is intended to make e2e tests accessible to all k8s developers since minikube runs across many platforms.
With this PR a dev workflow will be as follows:
0. Install minikube
1. Make changes to k8s HEAD
2. If the changes are not cloud provider specific, then run `make test-e2e-minikube`

In addition to minikube based e2e, this PR also adds a new makerule to run cross platform builds using docker - `make docker-cross`

As part of enabling minikube based e2e testing, this PR also introduces a new binary called `localkube` which is an implementation detail of minikube - think of kube master in a single process.

This PR is yet to be tested across all OSes. Validation help will be appreciated.

cc @dlorenc @pwittrock @kubernetes/sig-contributor-experience-misc

### CODE REVIEW

**Comment by fejta** [comment]
@k8s-bot test this

https://github.com/kubernetes/test-infra/pull/1486

File: hack/verify-flags/known-flags.txt [hunk 1]
```diff
@@ -635,3 +635,13 @@ garbage-collector-enabled
 viper-config
 log-lines-total
 run-duration
+apiserver-address
```
Comment (pwittrock) [inline hack/verify-flags/known-flags.txt]: Looks like most of the flags are sorted alphabetically.

File: Makefile [hunk 2]
```diff
@@ -194,6 +194,26 @@ test-e2e: ginkgo generated_files
...
+#   make test-e2e-minikube GINKGO_FLAGS="-ginkgo.focus=foo -ginkgo.skip=bar"
+endef
+.PHONY: test-e2e-minikube
+ifeq ($(PRINT_HELP),y)
+@echo "$$TEST_E2E_MINIKUBE_HELP_INFO"
+else
+test-e2e-minikube: ginkgo generated_files
+	WHAT=cmd/localkube hack/make-rules/docker-cross.sh
+	hack/make-rules/minikube-e2e-start.sh
+	KUBERNETES_CONFORMANCE_TEST=y hack/ginkgo-e2e.sh ${GINKGO_FLAGS}
```
Comment (pwittrock) [inline Makefile]: Do these run against local host by default?

File: hack/make-rules/minikube-e2e-start.sh [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# please refer to the kubernetes/release repo at
+# https://github.com/kubernetes/release.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_ROOT=$(dirname "${BASH_SOURCE}")/../..
+
+minikube start --kubernetes-version="file://${KUBE_ROOT}/_output/dockerized/bin/linux/amd64/localkube"
```
Comment (pwittrock) [inline hack/make-rules/minikube-e2e-start.sh]: Should we check that minikube is installed and at the correct version and provide a helpful error message if not?

**Comment by pwittrock** [comment]
@vishh Sorry about taking so long on the review.  Added a few comments, nothing major though.

**Comment by vishh** [comment]
@pwittrock no problem at all! It took me a while to get the tests to pass anyways.

**Comment by pwittrock** [comment]
SGTM.  Send a PTAL when you are ready for me to take another look.

File: hack/lib/golang.sh [hunk 4]
```diff
@@ -144,6 +144,7 @@ kube::golang::test_targets() {
     federation/cmd/genfeddocs
     vendor/github.com/onsi/ginkgo/ginkgo
     test/e2e/e2e.test
+    cmd/localkube
```
Comment (vishh) [inline hack/lib/golang.sh]: Sure. Moving it.

File: hack/make-rules/cross.sh [hunk 5]
```diff
@@ -27,6 +27,11 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
 # NOTE: Using "${array[*]}" here is correct.  [@] becomes distinct words (in
```
Comment (vishh) [inline hack/make-rules/cross.sh]: Done

File: hack/make-rules/cross.sh [hunk 6]
```diff
@@ -27,6 +27,11 @@ source "${KUBE_ROOT}/hack/lib/init.sh"
...
 # bash parlance).
 
+if [[ -n "${WHAT-}" ]]; then
```
Comment (vishh) [inline hack/make-rules/cross.sh]: Done

File: hack/verify-flags/known-flags.txt (hunk 1 above)
Comment (vishh) [inline hack/verify-flags/known-flags.txt]: Done

File: Makefile (hunk 2 above)
Comment (vishh) [inline Makefile]: Yeah. These tests are meant to be portable.

File: hack/make-rules/minikube-e2e-start.sh (hunk 3 above)
Comment (vishh) [inline hack/make-rules/minikube-e2e-start.sh]: Done!

**Comment by vishh** [comment]
@dlorenc I can use some help in resolving RBAC issues with localkube. This PR is currently broken due to recent changes to kube-apiserver and I don't have time now to fix localkube.

**Comment by dlorenc** [comment]
@vishh looking at this now. I was able to pull in your PR branch and build everything. What was failing for you?

**Comment by vishh** [comment]
apiserver doesn't come up. Try running `make test-e2e-minikube` against
this branch.

**Comment by dlorenc** [comment]
Did some debugging, it looks like the API server is actually coming up but having trouble. I'm still digging in to see if I can find the root cause. Curling the server over the insecure port works for some API requests, so it's up and listening:

```shell
$ wget -O - http://localhost:8080 2>/dev/null
{
  "paths": [
    "/api",
    "/api/v1",
    "/apis",
    "/apis/apps",
    "/apis/apps/v1beta1",
    "/apis/authentication.k8s.io",
    "/apis/authentication.k8s.io/v1beta1",
    "/apis/authorization.k8s.io",
    "/apis/authorization.k8s.io/v1beta1",
    "/apis/autoscaling",
    "/apis/autoscaling/v1",
    "/apis/batch",
    "/apis/batch/v1",
    "/apis/batch/v2alpha1",
    "/apis/certificates.k8s.io",
    "/apis/certificates.k8s.io/v1alpha1",
    "/apis/extensions",
    "/apis/extensions/v1beta1",
    "/apis/policy",
    "/apis/policy/v1beta1",
    "/apis/rbac.authorization.k8s.io",
    "/apis/rbac.authorization.k8s.io/v1alpha1",
    "/apis/storage.k8s.io",
    "/apis/storage.k8s.io/v1beta1",
    "/healthz",
    "/healthz/ping",
    "/healthz/poststarthook/bootstrap-controller",
    "/healthz/poststarthook/extensions/third-party-resources",
    "/healthz/poststarthook/rbac/bootstrap-roles",
    "/logs",
    "/metrics",
    "/swaggerapi/",
    "/ui/",
    "/version"
  ]
}
```

but `$ wget -O - http://localhost:8080/api/v1/pods 2>/dev/null` fails.

Not sure if this is related, but it's failing to register the node:
Jan 17 19:48:56 minikube localkube[6837]: E0117 19:48:56.109934    6837 kubelet_node_status.go:101] Unable to register node "minikube" with API server: rpc error: code = 13 desc = transport is closing

**Comment by vishh** [comment]
The last time I looked, RBAC webhook registration was failing.

**Comment by dlorenc** [comment]
Not sure how to push my commit here, but I think I have this fixed. Can you try the change in this branch? 

https://github.com/kubernetes/kubernetes/commit/27e1252cab7418d11a8ef75ec7339b5639448057

**Comment by vishh** [comment]
@dlorenc yeah. That does fix it!! Thanks!

**Comment by vishh** [comment]
Next step is to setup a k8s e2e test suite for this. @dlorenc will your team be able to own the test?

**Comment by pwittrock** [comment]
@vishh Do you need me to do any more review work on this?

**Comment by vishh** [comment]
@pwittrock Excepting the final commit, I don't think there is anything else to review.

**Comment by feiskyer** [comment]
Cool. It helps a lot for non-linux platforms.

**Review by dlorenc — COMMENTED** [review]
This looks good to me at this point. Not sure what other work is required to finish the PR, but let me know if I can help.

**Comment by vishh** [comment]
For starters, quite a few e2es are failing. We can get this PR merged and
then tackle the e2es. We need all green runs before publicizing this
feature. We need to add a test suite for running `make test-e2e-minikube`
on every PR.

**Pull request closed without merging**